scenario/state.py

```diff
@@ -165,67 +195,184 @@ class Relation(_DCBase):
         )
 
     @property
-    def joined_event(self):
+    def joined_event(self) -> "Event":
         """Sugar to generate a <this relation>-relation-joined event."""
         return Event(
             name=normalize_name(self.endpoint + "-relation-joined"), relation=self
         )
 
     @property
-    def created_event(self):
+    def created_event(self) -> "Event":
         """Sugar to generate a <this relation>-relation-created event."""
         return Event(
             name=normalize_name(self.endpoint + "-relation-created"), relation=self
         )
 
     @property
-    def departed_event(self):
+    def departed_event(self) -> "Event":
         """Sugar to generate a <this relation>-relation-departed event."""
         return Event(
             name=normalize_name(self.endpoint + "-relation-departed"), relation=self
         )
 
     @property
-    def broken_event(self):
+    def broken_event(self) -> "Event":
         """Sugar to generate a <this relation>-relation-broken event."""
         return Event(
             name=normalize_name(self.endpoint + "-relation-broken"), relation=self
         )
 
 
+def unify_ids_and_remote_units_data(ids: List[int], data: Dict[int, Any]):
+    """Unify and validate a list of unit IDs and a mapping from said ids to databag contents.
+
+    This allows the user to pass equivalently:
+    ids = []
+    data = {1: {}}
+
+    or
+
+    ids = [1]
+    data = {}
+
+    or
+
+    ids = [1]
+    data = {1: {}}
+
+    but catch the inconsistent:
+
+    ids = [1]
+    data = {2: {}}
+
+    or
+
+    ids = [2]
+    data = {1: {}}
+    """
+    if ids and data:
+        if not set(ids) == set(data):
+            raise StateValidationError(
+                f"{ids} should include any and all IDs from {data}"
+            )
+    elif ids:
+        data = {x: {} for x in ids}
+    elif data:
+        ids = [x for x in data]
+    else:
+        ids = [0]
+        data = {0: {}}
+    return ids, data
+
+
+@dataclasses.dataclass
+class Relation(RelationBase):
+    __type__ = RelationType.regular
+    remote_app_name: str = "remote"
+    remote_unit_ids: List[int] = dataclasses.field(default_factory=list)
+
+    # local limit
+    limit: int = 1
+
+    remote_app_data: Dict[str, str] = dataclasses.field(default_factory=dict)
+    remote_units_data: Dict[int, Dict[str, str]] = dataclasses.field(
+        default_factory=dict
+    )
+
+    @property
+    def __databags__(self):
+        yield self.local_app_data
+        yield self.local_unit_data
+        yield self.remote_app_data
+        yield from self.remote_units_data.values()
+
+    def __post_init__(self):
+        super().__post_init__()
+        self.remote_unit_ids, self.remote_units_data = unify_ids_and_remote_units_data(
+            self.remote_unit_ids, self.remote_units_data
+        )
+
+
+@dataclasses.dataclass
+class SubordinateRelation(RelationBase):
+    __type__ = RelationType.subordinate
```

Comment:
- Why dunder instead of `_type` or `__type`?
- Why not `TYPE`?
- Why is it needed if we have `isinstance`?

Reply:
- no specific reason, I like dunders. They look pythonic
- wanted to make it obvious that it's not a user-facing variable (it being a dataclass and all, I was afraid people might think they can supply it as an argument)
- because cyclic imports (and I was too lazy to fix them by moving the abc into a separate file or something else, and it felt ugly to do lazy imports).

Reply:
do you have a better solution in mind? or should I disentangle the cyclic imports?